src/shared/mkfs-util.c

```diff
@@ -102,6 +103,11 @@ int make_filesystem(
         assert(fstype);
         assert(label);
 
+        if (fstype_is_ro(fstype) && !root)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Cannot generate read-only filesystem %s without a source tree.",
+                                       fstype);
+
```

Comment:
hmm, for the read-only file systems we really need explicit support, we cannot have generic fallback support as we have for the writable fs (where we at a minimum can call `mkfs.<fstype>` without extra arguments to create the fs).

sounds like something to integrate into the large block of ifs below, i.e.:

```c
if (streq(fstype, "swap")) {
        …
} else if (streq(fstype, "squashfs")) {
        …
} else if (fstype_is_ro(fstype)) 
        return log_error_ernro(SYNTHETIC_ERRNO(EOPNOTSUP), "Don't know how to create read-only file system '%s', refusing.", fstype);
else {
        … 
}
``